### PR TITLE
STR-534: Bump RETH version to 1.1.0 and all ALLOY accordingly

### PR DESCRIPTION
**Pull request by evgenyzdanovich**
## Description

Bump RETH version to 1.1.0 and all ALLOY crates accordingly as reth depends on it. 
Adjust the codebase following the new reth version.

### Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update
-   [x] Refactor
-   [ ] New or updated tests

## Checklist

-   [x] I have performed a self-review of my code.
-   [x] I have commented my code where necessary.
-   [ ] I have updated the documentation if needed.
-   [ ] My changes do not introduce new warnings.
-   [x] I have added tests that prove my changes are effective or that my feature works.
-   [x] New and existing tests pass with my changes.

### CODE REVIEW

**Review by delbonis — APPROVED** [review]
Some of the prover changes might have to be redone due to #478, depending on how things get merged.

**Comment by delbonis** [comment]
Oops meant to leave that review as just a comment instead of an approval since this is just a draft.

**Review by evgenyzdanovich — COMMENTED** [review]
PTAL

File: Cargo.toml [hunk 1]
```diff
@@ -111,20 +111,83 @@ strata-tx-parser = { path = "crates/tx-parser" }
...
 base64 = "0.22.1"
-bincode = "1.3.3"
+bincode = "1.3"
 bitcoin = { version = "=0.32.3", features = ["serde"] }
 bitcoind = { version = "0.36.0", features = ["26_0"] }
 borsh = { version = "1.5.0", features = ["derive"] }
 bytes = "1.6.0"
 chrono = "0.4.38"
+clap = "4"
+derive_more = { version = "1", features = ["full"] }
```
Comment (storopoli) [inline Cargo.toml:190]: We are not bringing `clap` to the codebase. Please use `argh` since it is less compile time and we are already using it throughout the codebase.

Comment (storopoli) [inline Cargo.toml:190]: Why do we need `derive_more`?

File: bin/strata-reth/Cargo.toml [hunk 2]
```diff
@@ -8,26 +8,31 @@ name = "strata-reth"
...
+strata-reth-db.workspace = true
+strata-reth-exex.workspace = true
+strata-reth-node.workspace = true
+strata-reth-rpc.workspace = true
+
+alloy-genesis.workspace = true
+alloy-rpc-types.workspace = true
 async-trait.workspace = true
-clap = "4"
+clap.workspace = true
```
Comment (storopoli) [inline bin/strata-reth/Cargo.toml:19]: Oh, that's why we have `clap`.
That's annoying :/ Gonna open a ticket for that

Comment (storopoli) [reply bin/strata-reth/Cargo.toml:19]: Reopening STR-267

File: bin/strata-reth/src/main.rs [hunk 3]
```diff
@@ -107,6 +106,20 @@ pub struct AdditionalConfig {
...
+
+impl ChainSpecParser for StrataChainSpecParser {
+    type ChainSpec = ChainSpec;
+
+    const SUPPORTED_CHAINS: &'static [&'static str] = &["dev", "devnet"];
+
+    fn parse(s: &str) -> eyre::Result<Arc<Self::ChainSpec>> {
+        chain_value_parser(s)
+    }
+}
```
Comment (storopoli) [inline bin/strata-reth/src/main.rs:121]: I think this should be an `enum` with the `#[non_exhaustive]` and the `SUPPORTED_CHAINS` will be the variants.
This makes it easier to catch in `semver-checks` in the future.

File: crates/primitives/src/l1.rs [hunk 4]
```diff
@@ -21,6 +21,7 @@ use bitcoin::{
 use borsh::{BorshDeserialize, BorshSerialize};
 use rand::rngs::OsRng;
 use reth_primitives::revm_primitives::FixedBytes;
+use secp256k1::SECP256K1;
```
Comment (storopoli) [inline crates/primitives/src/l1.rs:24]: I prefer to keep it as `bitcoin::secp256k1::SECP256K1` as before since it is easier to keep in-sync with `bitcoin`.
Can we revert this change?

File: crates/reth/node/src/evm.rs [hunk 5]
```diff
@@ -1,43 +1,48 @@
...
+    },
+    Header, TransactionSigned,
+};
 use revm::{inspector_handle_register, Database, Evm, EvmBuilder, GetInspector};
-use revm_primitives::{Address, AnalysisKind, Bytes, CfgEnvWithHandlerCfg, Env, TxEnv, U256};
 use strata_reth_evm::set_evm_handles;
 
 /// Custom EVM configuration
-#[derive(Debug, Clone, Copy, Default)]
+#[derive(Debug, Clone)]
```
Comment (storopoli) [inline crates/reth/node/src/evm.rs:16]: Why are we removing `Copy`?

File: crates/reth/exex/src/alloy2reth.rs [hunk 6]
```diff
@@ -1,6 +1,5 @@
 use alloy_rpc_types::Withdrawal as AlloyWithdrawal;
 use reth_primitives::Withdrawal as RethWithdrawal;
-
```
Comment (storopoli) [inline crates/reth/exex/src/alloy2reth.rs:3]: Don't delete this empty space.
Or is this `cargo fmt`?

File: crates/reth/exex/src/prover_exex.rs [hunk 7]
```diff
@@ -1,18 +1,20 @@
 use std::{collections::HashMap, sync::Arc};
 
+// use reth_execution_types::BlockExecutionInput;
```
Comment (storopoli) [inline crates/reth/exex/src/prover_exex.rs:3]: Don't comment it out. Please remove it

File: crates/reth/node/src/validator.rs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    validate_version_specific_fields, EngineApiMessageVersion, EngineObjectValidationError,
+    EngineTypes, EngineValidator, FullNodeTypes, NodeTypesWithEngine, PayloadOrAttributes,
+};
+
+use crate::{StrataEngineTypes, StrataPayloadAttributes};
+
+/// Strata engine validator
+#[derive(Debug, Clone)]
+pub struct StrataEngineValidator {
+    chain_spec: Arc<ChainSpec>,
```
Comment (storopoli) [inline crates/reth/node/src/validator.rs:15]: We don't need to `Arc` it.
Just say 
```suggestion
    chain_spec: &ChainSpec,
```

Then the implementer deals with with `Arc`ing or not.

Example https://github.com/alpenlabs/strata/blob/f40ecce865a134511f0dc7b70c08eecce7449551/crates/consensus-logic/src/duty/worker.rs#L112

**Review by storopoli — CHANGES_REQUESTED** [review]

**Review by evgenyzdanovich — COMMENTED** [review]
Thanks for the timely review @storopoli , PTAL!

File: crates/reth/node/src/evm.rs (hunk 5 above)
Comment (storopoli) [reply crates/reth/node/src/evm.rs:16]: Ok

File: crates/util/python-utils/src/constants.rs [hunk 9]
```diff
@@ -2,9 +2,10 @@ use std::sync::LazyLock;
 
 use bdk_wallet::bitcoin::{
     key::Parity,
-    secp256k1::{PublicKey, SecretKey, SECP256K1},
+    secp256k1::{PublicKey, SecretKey},
     Amount, Network, XOnlyPublicKey,
 };
```
Comment (storopoli) [inline crates/util/python-utils/src/constants.rs:7]: Again, move `SECP256K1` to `bitcoin::secp256k1`

File: crates/util/python-utils/src/taproot.rs [hunk 10]
```diff
@@ -6,15 +6,16 @@ use bdk_bitcoind_rpc::{
...
+        consensus::serialize, key::Parity, Address, AddressType, FeeRate, PublicKey, Transaction,
+        XOnlyPublicKey,
     },
     descriptor::IntoWalletDescriptor,
     miniscript::ToPublicKey,
     KeychainKind, Wallet,
 };
 use musig2::KeyAggContext;
 use pyo3::prelude::*;
+use secp256k1::SECP256K1;
```
Comment (storopoli) [inline crates/util/python-utils/src/taproot.rs:18]: Same as above

**Review by storopoli — CHANGES_REQUESTED** [review]

File: bin/strata-reth/Cargo.toml (hunk 2 above)
Comment (storopoli) [reply bin/strata-reth/Cargo.toml:19]: Sure keeping tracking in STR-267.
Cc @delbonis because of the `argh` history.

File: crates/reth/node/src/validator.rs (hunk 8 above)
Comment (storopoli) [reply crates/reth/node/src/validator.rs]: Ok

File: crates/util/python-utils/Cargo.toml [hunk 11]
```diff
@@ -21,7 +21,7 @@ bdk_wallet = "1.0.0-beta.5"
 # "abi3-py310" tells pyo3 (and maturin) to build using the stable ABI with minimum Python version 3.10
 musig2.workspace = true
 pyo3 = { version = "0.22.6", features = ["extension-module", "abi3-py310"] }
-secp256k1 = { workspace = true, features = ["rand-std"] }
+secp256k1 = { workspace = true, features = ["rand-std", "global-context"] }
```
Comment (storopoli) [inline crates/util/python-utils/Cargo.toml:24]: We don't need this because we have `bdk_wallet::bitcoin::secp256k1` re-exported from `bdk`'s `bitcoin` and then `bitcoin`'s `secp256k1`

**Review by storopoli — CHANGES_REQUESTED** [review]
Final nit. This is annoying since `scp256k1` might become out of sync with `bdk_wallet::bitcoin`.

File: crates/util/python-utils/Cargo.toml (hunk 11 above)
Comment (evgenyzdanovich) [reply crates/util/python-utils/Cargo.toml]: Yes, I understand this and this is what surprises me. But the `python-utils` strangely stops compiling if I remove this as a _direct_ dependency (**even though it's not used it as a DIRECT import**, but through the chain of re-imports).

Even though I understand it's not ideal, I propose for the middleground: I'll leave TODO in `python-utils/cargo.toml` to get rid of this and we can figure it out later (ideally together, please)?

**Review by delbonis — DISMISSED** [review]

File: crates/util/python-utils/Cargo.toml (hunk 11 above)
Comment (storopoli) [reply crates/util/python-utils/Cargo.toml]: Ok we can debug this later.
To be honest, I don't want to have to use Python for func-tests.
So in the future all of this crate's code would be migrated over to a Rust func-test harness somehow.

**Review by storopoli — APPROVED** [review]
ACK a41e619a216745f94080c583fe5745ab1c312b4a

**Review by storopoli — DISMISSED** [review]
EDIT: I was wrong reverted changes back.

**Review by storopoli — APPROVED** [review]
ACK ce1299d10725ec18078ddd8ad4f44aaca603c452

**Review by prajwolrg — APPROVED** [review]
ACK https://github.com/alpenlabs/strata/commit/ce1299d10725ec18078ddd8ad4f44aaca603c452

File: crates/proof-impl/evm-ee-stf/src/db.rs [hunk 12]
```diff
@@ -162,7 +172,7 @@ impl InMemoryDBHelper for InMemoryDB {
     }
 
     fn storage_keys(&self) -> HashMap<Address, Vec<U256>> {
-        let mut out = HashMap::new();
+        let mut out = HashMap::with_hasher(alloy_primitives::map::DefaultHashBuilder::default());
```
[the remark this answers is not in the document]
Comment (storopoli) [reply crates/proof-impl/evm-ee-stf/src/db.rs:173]: Can we resolve this?
